ortb/request_validator.go

```diff
@@ -0,0 +1,195 @@
+				prebid.Bidder[k] = v
+				prebidModified = true
+			}
+			delete(ext, k)
+			extModified = true
+		}
+	}
+
+	if hasStoredResponses && prebid.StoredAuctionResponse == nil {
+		return []error{fmt.Errorf("request validation failed. The StoredAuctionResponse.ID field must be completely present with, or completely absent from, all impressions in request. No StoredAuctionResponse data found for request.imp[%d].ext.prebid \n", impIndex)}
+	}
+
+	if err := rv.validateStoredBidResponses(prebid, storedBidResp, imp.ID); err != nil {
+		return []error{err}
+	}
+
+	errL := []error{}
+
+	for bidder, ext := range prebid.Bidder {
+		coreBidder, _ := openrtb_ext.NormalizeBidderName(bidder)
+		if tmp, isAlias := aliases[bidder]; isAlias {
+			coreBidder = openrtb_ext.BidderName(tmp)
+		}
+
+		if coreBidderNormalized, isValid := rv.bidderMap[coreBidder.String()]; isValid {
+			if err := rv.paramsValidator.Validate(coreBidderNormalized, ext); err != nil {
+				return []error{fmt.Errorf("request.imp[%d].ext.prebid.bidder.%s failed validation.\n%v", impIndex, bidder, err)}
+			}
+		} else {
+			if msg, isDisabled := rv.disabledBidders[bidder]; isDisabled {
+				errL = append(errL, &errortypes.BidderTemporarilyDisabled{Message: msg})
+				delete(prebid.Bidder, bidder)
+				prebidModified = true
+			} else {
+				return []error{fmt.Errorf("request.imp[%d].ext.prebid.bidder contains unknown bidder: %s. Did you forget an alias in request.ext.prebid.aliases?", impIndex, bidder)}
+			}
+		}
+	}
+
+	if len(prebid.Bidder) == 0 {
+		errL = append(errL, fmt.Errorf("request.imp[%d].ext.prebid.bidder must contain at least one bidder", impIndex))
+		return errL
+	}
+
+	if prebidModified {
+		impExt.SetPrebid(prebid)
+	}
+	if extModified {
+		impExt.SetExt(ext)
+	}
+
+	return errL
+}
+
+func (rv *RequestValidator) validateStoredBidResponses(prebid *openrtb_ext.ExtImpPrebid, storedBidResp stored_responses.ImpBidderStoredResp, impId string) error {
+	if storedBidResp == nil && len(prebid.StoredBidResponse) == 0 {
+		return nil
+	}
+
+	if storedBidResp == nil {
+		return generateStoredBidResponseValidationError(impId)
+	}
+	if bidResponses, ok := storedBidResp[impId]; ok {
+		if len(bidResponses) != len(prebid.Bidder) {
+			return generateStoredBidResponseValidationError(impId)
+		}
+
+		for bidderName := range bidResponses {
+			if _, bidderNameOk := openrtb_ext.NormalizeBidderName(bidderName); !bidderNameOk {
+				return fmt.Errorf(`unrecognized bidder "%v"`, bidderName)
+			}
+			if _, present := prebid.Bidder[bidderName]; !present {
+				return generateStoredBidResponseValidationError(impId)
+			}
+		}
+	}
+	return nil
+}
+
+func generateStoredBidResponseValidationError(impID string) error {
+	return fmt.Errorf("request validation failed. Stored bid responses are specified for imp %s. Bidders specified in imp.ext should match with bidders specified in imp.ext.prebid.storedbidresponse", impID)
+}
+
+// TODO: move this outside of this package
```

Comment:
Where should this function live? It was originally in `endpoints/openrtb2/auction.go` and was being used for both imp validation and in the function `mergeBidderParamsImpExt` which were both in the same package. For now I just copied the function here but this seems like an awkward place for it. Perhaps it belongs in some package other than `ortb` and `endpoints/openrtb2`?

Reply:
Discussed and can move into `bidders.go#IsPotentialBidder` below `IsBidderNameReserved`.